### PR TITLE
Store the whole clientDataJSON, but compressed

### PR DESCRIPTION
**Pull request by hrxi**
This makes reconstruction trivial. Previously, reconstruction of the original JSON was a delicate process, even more so since Android Chrome's behavior violates the standard: https://issues.chromium.org/issues/40191627.

Compressing is done based on expected data, this allows to save a lot of space for the redundant challenge information and the general JSON structure.

This is an alternative to #2227, also fixing https://github.com/nimiq/core-rs-albatross/pull/1867#discussion_r1470116959.

### CODE REVIEW

File: primitives/transaction/src/signature_proof.rs [hunk 1]
```diff
@@ -310,31 +286,10 @@ bitflags! {
...
-    pub struct WebauthnClientDataFlags: u8 {
-        const NO_CROSSORIGIN_FIELD  = 1 << 0;
-        const ESCAPED_ORIGIN_SLASHES = 1 << 1;
-
-        // const HAS_EXTRA_FIELDS = 1 << 7; // TODO Replace client_data_extra_fields length null byte when no extra fields are present
-    }
-}
-
 #[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
 pub struct WebauthnExtraFields {
```
Comment (jsdanielh) [inline primitives/transaction/src/signature_proof.rs:290]: Could you add some `rustdoc` to the struct and its members?

File: primitives/transaction/src/signature_proof.rs [hunk 2]
```diff
@@ -154,86 +155,61 @@ impl SignatureProof {
...
-            "https"
+        let challenge: Blake2bHash = message.hash();
+
+        let client_data_json = match decompress(
+            &predicted(challenge.as_slice()),
+            &webauthn_fields.client_data_json_compressed,
+        ) {
+            Ok(client_data_json) => client_data_json,
+            Err(e) => {
+                debug!("Failed to decompress client data JSON: {e}");
```
Comment (jsdanielh) [inline primitives/transaction/src/signature_proof.rs:167]:
```suggestion
                debug!(error=e, "Failed to decompress client data JSON");
```

File: primitives/transaction/src/signature_proof.rs [hunk 3]
```diff
@@ -154,86 +155,61 @@ impl SignatureProof {
...
-            .client_data_flags
-            .contains(WebauthnClientDataFlags::ESCAPED_ORIGIN_SLASHES)
-        {
-            r":\/\/"
-        } else {
-            "://"
+        let client_data: WebauthnClientData = match serde_json::from_slice(&client_data_json) {
+            Ok(client_data) => client_data,
+            Err(e) => {
+                debug!("Failed to read client data JSON: {e}");
```
Comment (jsdanielh) [inline primitives/transaction/src/signature_proof.rs:175]:
```suggestion
                debug!(error=e, "Failed to read client data JSON");
```

File: primitives/transaction/src/signature_proof.rs [hunk 4]
```diff
@@ -154,86 +155,61 @@ impl SignatureProof {
...
+        };
 
-        let hostname = url.host_str();
-        if hostname.is_none() {
-            debug!("Failed to extract hostname: {:?}", url);
+        let challenge_base64 = BASE64_URL_SAFE_NO_PAD.encode(challenge);
+        if challenge_base64 != client_data.challenge {
+            debug!(
+                "challenge mismatch: expected {}, got {}",
+                client_data.challenge, challenge_base64,
```
Comment (jsdanielh) [inline primitives/transaction/src/signature_proof.rs:192]:
```suggestion
                expected=client_data.challenge, obtained=challenge_base64,
                "challenge mismatch",
```

**Review by jsdanielh — COMMENTED** [review]
Just some small suggestions

**Comment by hrxi** [comment]
Superseded by #2275.

**Pull request closed without merging**